src/cloudformation_cli_python_lib/recast.py

```diff
@@ -0,0 +1,131 @@
+    if isinstance(item, dict):
+        recast_object(cls, item, classes)
+        return item
+    raise InvalidRequest(f"Unsupported type: {type(item)} for {k}")
+
+
+def _recast_primitive(cls: Any, k: str, v: str) -> Any:
+    if cls == typing.Any:
+        # If the type is Any, we cannot guess what the original type was, so we leave
+        # it as a string
+        return v
+    if cls == bool:
+        if v.lower() == "true":
+            return True
+        if v.lower() == "false":
+            return False
+        raise InvalidRequest(f'value for {k} "{v}" is not boolean')
+    return cls(v)
+
+
+# yes, introspecting type hints is ugly, but hopefully only needed temporarily
+def _field_to_type(field: Any, key: str, classes: Dict[str, Any]) -> Any:  # noqa: C901
+    if field in [int, float, str, bool, typing.Any]:
+        return field
+    # If it's a ForwardRef we need to find base type
+    if isinstance(field, get_forward_ref_type()):
+        # Assuming codegen added an _ as a prefix, removing it and then getting the
+        # class from model classes
+        return classes[field.__forward_arg__[1:]]
+    # Assuming this is a generic object created by typing.Union
+    try:
+        possible_types = field.__args__
+        if not possible_types:
+            raise InvalidRequest(f"Cannot process type {field} for field {key}")
+    except AttributeError:
+        raise InvalidRequest(f"Cannot process type {field} for field {key}")
+    # Assuming that the union is generated from typing.Optional, so only
+    # contains one type and None
+    # pylint: disable=unidiomatic-typecheck
+    fields = [t for t in possible_types if type(None) != t]
+    if len(fields) != 1:
+        raise InvalidRequest(f"Cannot process type {field} for field {key}")
+    field = fields[0]
+    # If it's a primitive we're done
+    if field in [int, float, str, bool, typing.Any]:
+        return field
+    # If it's a ForwardRef we need to find base type
+    if isinstance(field, get_forward_ref_type()):
+        # Assuming codegen added an _ as a prefix, removing it and then getting the
+        # class from model classes
+        return classes[field.__forward_arg__[1:]]
+    # reduce Sequence/AbstractSet to inner type
+    if str(field).startswith("typing.Sequence") or str(field).startswith(
+        "typing.AbstractSet"
+    ):
+        return _field_to_type(field.__args__[0], key, classes)
+    if str(field).startswith("typing.MutableMapping"):
+        return _field_to_type(field.__args__[1], key, classes)
+    # If it's a type we don't know how to handle, we bail
+    raise InvalidRequest(f"Cannot process type {field} for field {key}")
+
+
+# pylint: disable=protected-access,no-member
+def get_forward_ref_type() -> Any:
+    # ignoring mypy on the import as it catches (_)ForwardRef as invalid, use for
+    # introspection is valid:
+    # https://docs.python.org/3/library/typing.html#typing.ForwardRef
+    if "ForwardRef" in dir(typing):
```

Comment:
whats going on here? in what cases are we using the protected _ForwardRef

Reply:
in python 3.6 it's private, was made public in 3.7